webgoat-lessons/jwt/src/main/java/org/owasp/webgoat/jwt/JWTFinalEndpoint.java

```diff
@@ -0,0 +1,195 @@
+import io.jsonwebtoken.impl.TextCodec;
+
+/**
+ * <pre>
+ *  {
+ *      "typ": "JWT",
+ *      "kid": "webgoat_key",
+ *      "alg": "HS256"
+ *  }
+ *  {
+ *       "iss": "WebGoat Token Builder",
+ *       "iat": 1524210904,
+ *       "exp": 1618905304,
+ *       "aud": "webgoat.org",
+ *       "sub": "jerry@webgoat.com",
+ *       "username": "Jerry",
+ *       "Email": "jerry@webgoat.com",
+ *       "Role": [
+ *       "Cat"
+ *       ]
+ *  }
+ * </pre>
+ *
+ * @author nbaars
+ * @since 4/23/17.
+ */
+@RestController
+@AssignmentHints({"jwt-final-hint1", "jwt-final-hint2", "jwt-final-hint3", "jwt-final-hint4", "jwt-final-hint5", "jwt-final-hint6"})
+public class JWTFinalEndpoint extends AssignmentEndpoint {
+
+    private final DataSource dataSource;
+
+    private JWTFinalEndpoint(DataSource dataSource) {
+        this.dataSource = dataSource;
+    }
+
+    @PostMapping(path="/JWT/encode",produces=MediaType.TEXT_HTML_VALUE)
+    @ResponseBody
+    public String encode(@RequestParam("jsonHeader") String jsonHeader,
+    		@RequestParam("jsonPayload") String jsonPayload,
+    		@RequestParam("jsonSecret") String jsonSecret) throws NoSuchAlgorithmException {		
+		
+    	//System.out.println(jsonHeader);
+    	//System.out.println(jsonPayload);
+		String encodedHeader;
+		String encodedPayload;
+		String encodedSignature;
+		try {
+			encodedHeader = TextCodec.BASE64URL.encode(jsonHeader);
+			encodedPayload = TextCodec.BASE64URL.encode(jsonPayload);
+			if (jsonHeader.toLowerCase().contains("none")) {
+				encodedSignature="";
+			} else {
+				encodedSignature = TextCodec.BASE64URL.encode(getJWTSignature(jsonHeader, encodedHeader, encodedPayload, jsonSecret));
+			}
+		} catch (Exception e) {
+			encodedHeader="";
+			encodedPayload="signature type not supported in this tool, try jwt.io";
+			encodedSignature = "";
+		}
+		String result = "{\"encodedHeader\":\""+encodedHeader+"\",\"encodedPayload\":\""+encodedPayload+"\",\"encodedSignature\":\""+encodedSignature+"\"}";
+		//System.out.println(result);
+		return result;
+    }
+       
+    private byte[] getJWTSignature(String jsonHeader, String encodedHeader, String encodedPayload, String jsonSecret) throws NoSuchAlgorithmException, InvalidKeyException {
+    	String message = encodedHeader+"."+encodedPayload;
+  	  	String algorithm = "HmacSHA256";
+    	if (jsonHeader.equals("HS512")) {
+    		algorithm = "HmacSHA512";
+    	} 
+    	Mac macInstance = Mac.getInstance(algorithm);
+        SecretKeySpec secret_key = new SecretKeySpec(TextCodec.BASE64.decode(jsonSecret), algorithm);
+        macInstance.init(secret_key);
+
+        return macInstance.doFinal(message.getBytes(StandardCharsets.UTF_8));
+    }
+    
+    @PostMapping(path="/JWT/decode",produces=MediaType.TEXT_HTML_VALUE)
+    @ResponseBody
+    public String decode(@RequestParam("jwtToken") String jwtToken) throws NoSuchAlgorithmException {		
+		try {
+			String encodedHeader = jwtToken.substring(0, jwtToken.indexOf("."));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

If the indexOf() search on line 134 does not find the substring and returns -1, the call to substring() on line 134 will throw IndexOutOfBoundsException. Note that the method is publicly accessible to external callers who might not be aware of the constraint on string jwtToken.